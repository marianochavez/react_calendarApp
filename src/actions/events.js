import Swal from "sweetalert2";
import { fetchWithToken } from "../helpers/fetch";
import { prepareEvents } from "../helpers/prepareEvents";
import { types } from "../types/types";

export const eventStartAddNew = (event) => {
    return async(dispatch, getState) => {

        const { uid, name } = getState().auth;

        try {
            const res = await fetchWithToken(`event`, event, 'POST');
            const body = await res.json();

            if ( body.ok ) {
                event.id = body.eventSaved.id;
                event.user = {
                    _id: uid,
                    name: name
                }
                dispatch(eventAddNew( event ));
            }
        } catch (error) {
            console.log(error);
        }
        
    }
}

const eventAddNew = (event) => ({
    type: types.eventAddNew,
    payload: event
});

export const eventSetActive = (event) => ({
    type: types.eventSetActive,
    payload: event
});

export const eventClearActiveEvent = () => ({
    type: types.eventClearActiveEvent
});

export const eventStartUpdate = (event) => {
    return async(dispatch) => {
        try {
            const res = await fetchWithToken(`event/${event.id}`, event, 'PUT');
            const body = await res.json();

            if ( body.ok ) {
                dispatch(eventUpdated( event ));
            } else {
                Swal.fire('Error', body.msg, 'error');
            }
        } catch (error) {
            console.log(error);
        }
    }
};

const eventUpdated = (event) => ({
    type: types.eventUpdated,
    payload: event
});

export const eventStartDelete = () => {
    return async(dispatch, getState) => {

        const { id } = getState().calendar.activeEvent;

        try {
            const res = await fetchWithToken(`event/${id}`, {}, 'DELETE');
            const body = await res.json();
            
            if ( body.ok ) {
                dispatch(eventDeleted());
            } else {
                Swal.fire('Error', body.msg, 'error');
            }
        } catch (error) {
            console.log(error);
        }
    }
};

const eventDeleted = () => ({
    type: types.eventDeleted,
});

export const eventStartLoading = () => {
    return async(dispatch) => {
        
        try {
            
            const res = await fetchWithToken('event');
            const body = await res.json();
            const events = prepareEvents(body.events);

            if ( body.ok ) {
                dispatch(eventLoaded( events ));
            }

        } catch (error) {
            console.log(error);
        }
    }
};

const eventLoaded = (events) => ({
    type: types.eventLoaded,
    payload: events
});

export const eventLogout = () => ({type: types.eventLogout});