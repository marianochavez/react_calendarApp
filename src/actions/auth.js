import Swal from "sweetalert2";

import { eventLogout } from "./events";
import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch";
import { types } from "../types/types";

export const startLogin = (email, password) => {
  return async (dispatch) => {
    const res = await fetchWithoutToken(
      "auth", // endpoint
      { email, password },
      "POST"
    );

    const body = await res.json();

    if (body.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(
        login({
          uid: body.uid,
          name: body.name,
        })
      );
    } else {
      Swal.fire("Error", body.msg, "error");
    }
  };
};

export const startRegister = (name, email, password) => {
  return async (dispatch) => {
    const res = await fetchWithoutToken(
      "auth/new", // endpoint
      { name, email, password },
      "POST"
    );

    const body = await res.json();

    if (body.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(
        login({
          uid: body.uid,
          name: body.name,
        })
      );
    } else {
      Swal.fire("Error", body.msg, "error");
    }
  };
};

export const startChecking = () => {
  return async (dispatch) => {
    dispatch(checkingStart());
    const res = await fetchWithToken("auth/renew");
    const body = await res.json();

    if (body.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());

      dispatch(
        login({
          uid: body.uid,
          name: body.name,
        })
      );
    } else {
      dispatch(checkingFinish());
    }
  };
};

const checkingFinish = () => ({
    type: types.authCheckingFinish,
});

const checkingStart = () => ({
    type: types.authCheckingStart,
});

const login = (user) => ({
  type: types.authLogin,
  payload: user,
});

export const startLogout = () => {
  return async (dispatch) => {
    localStorage.clear();
    dispatch(eventLogout());
    dispatch(logout());
  };
};

const logout = () => ({
  type: types.authLogout,
});

