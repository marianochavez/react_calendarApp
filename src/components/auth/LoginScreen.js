import React from "react";
import { useDispatch } from "react-redux";
import { Link } from "react-router-dom";
import { startLogin } from "../../actions/auth";

import { useForm } from "../../hooks/useForm";
import "./login.css";

export const LoginScreen = () => {

  const dispatch = useDispatch();

  const [ formLoginValues, handleLoginInputChange ] = useForm({
    lEmail: "mariano@gmail.com",
    lPassword: "123456",
  });

  const { lEmail, lPassword } = formLoginValues;

  const handleLogin = (e) => {
    e.preventDefault();
    dispatch(startLogin(lEmail, lPassword));
  };

  return (
    <section className="vh-100 gradient-custom">
      <div className="container py-5 h-100">
        <div className="row d-flex justify-content-center align-items-center h-100">
          <div className="col-12 col-md-8 col-lg-6 col-xl-5">
            <div className="card card-login bg-dark text-white animate__animated animate__fadeIn animate__fast">
              <div className="card-body p-5 text-center">
                <div className="mb-md-5 mt-md-4 pb-5">
                  <h2 className="fw-bold mb-2 text-uppercase">Login</h2>
                  <p className="text-white-50 mb-5">
                    Ingresar email y contraseña!
                  </p>
                  <form onSubmit={handleLogin}>
                    <div className="form-outline form-white mb-4">
                      <input
                        type="email"
                        id="typeEmailX"
                        placeholder="Email"
                        name="lEmail"
                        value={lEmail}
                        onChange={handleLoginInputChange}
                        autoComplete="off"
                        className="form-control form-control-lg"
                      />
                    </div>

                    <div className="form-outline form-white mb-4">
                      <input
                        autoComplete="off"
                        placeholder="Contraseña"
                        type="password"
                        name="lPassword"
                        value={lPassword}
                        onChange={handleLoginInputChange}
                        id="typePasswordX"
                        className="form-control form-control-lg"
                      />
                    </div>

                    <button
                      className="btn btn-outline-light btn-lg px-5"
                      type="submit"
                    >
                      Ingresar
                    </button>

                  </form>
                  
                </div>

                <div>
                  <p className="mb-0">
                    No tienes cuenta?{" "}
                    <Link to="/register" className="text-white-50 fw-bold">
                      Registrarme
                    </Link>
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </section>
  );
};
