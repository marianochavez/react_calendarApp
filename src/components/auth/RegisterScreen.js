import React from 'react'
import { useDispatch } from 'react-redux';
import { Link } from 'react-router-dom'
import Swal from 'sweetalert2';
import { startRegister } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';

export const RegisterScreen = () => {

  const dispatch = useDispatch();

  const [ formRegisterValues, handleRegisterInputChange ] = useForm({
    rName: "",
    rEmail: "",
    rPassword: "",
    rPassword2: "",
  });

  const { rName, rEmail, rPassword, rPassword2 } = formRegisterValues;

  const handleRegister = (e) => {
    e.preventDefault();

    if ( rPassword !== rPassword2 ) {
      return Swal.fire('Error', 'Las contraseñas no coinciden', 'error');
    }

    dispatch(startRegister(rName, rEmail, rPassword));
  };



  return (
    <section className="vh-100 gradient-custom">
      <div className="container py-5 h-100">
        <div className="row d-flex justify-content-center align-items-center h-100">
          <div className="col-12 col-md-8 col-lg-6 col-xl-5">
            <div className="card card-login bg-dark text-white animate__animated animate__fadeIn animate__fast">
              <div className="card-body p-5 text-center">
                <div className="mb-md-5 mt-md-4 pb-4">
                  <h2 className="fw-bold mb-2 text-uppercase">Registro</h2>
                  <p className="text-white-50 mb-5">
                    Por favor creese una cuenta!
                  </p>
                  <form onSubmit={handleRegister}>
                    
                    <div className="form-outline form-white mb-4">

                      <input
                        type="text"
                        autoComplete="off"
                        placeholder='Nombre'
                        name="rName"
                        value={rName}
                        onChange={handleRegisterInputChange}
                        className="form-control form-control-lg"
                      />
                    </div>

                    <div className="form-outline form-white mb-4">

                      <input
                        type="email"
                        autoComplete="off"
                        placeholder='Email'
                        name="rEmail"
                        value={rEmail}
                        onChange={handleRegisterInputChange}
                        className="form-control form-control-lg"
                      />
                    </div>

                    <div className="form-outline form-white mb-4">

                      <input
                        autoComplete="off"
                        type="password"
                        id="typePasswordX"
                        placeholder='Contraseña'
                        name="rPassword"
                        value={rPassword}
                        onChange={handleRegisterInputChange}
                        className="form-control form-control-lg"
                      />
                    </div>
                    
                    <div className="form-outline form-white mb-4">

                      <input
                        autoComplete="off"
                        type="password"
                        id="typePasswordX2"
                        placeholder='Confirmar contraseña'
                        name="rPassword2"
                        value={rPassword2}
                        onChange={handleRegisterInputChange}
                        className="form-control form-control-lg"
                      />
                    </div>

                    <button
                      className="btn btn-outline-light btn-lg px-5"
                      type="submit"
                    >
                      Registrarme
                    </button>

                  </form>

      
                </div>

                <div>
                  <p className="mb-0">
                    Ya estas registrado?{" "}
                    <Link to="/login" className="text-white-50 fw-bold">
                      Ingresar
                    </Link>
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </section>
  )
}
