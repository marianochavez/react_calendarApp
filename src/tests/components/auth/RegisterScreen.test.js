/** * @jest-environment node */
import 'jsdom-global/register';
import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import Swal from 'sweetalert2';

import '@testing-library/jest-dom';
import { startLogin, startRegister } from '../../../actions/auth';
import { MemoryRouter } from 'react-router-dom';
import { RegisterScreen } from '../../../components/auth/RegisterScreen';


jest.mock('../../../actions/auth', () => ({
    startRegister: jest.fn()
}));

jest.mock('sweetalert2', () => ({
    fire: jest.fn(),
}));

const middlewares = [ thunk ];
const mockStore = configureStore( middlewares );

const initState = {};
const store = mockStore( initState );
store.dispatch = jest.fn();


const wrapper = mount(
    <Provider store={ store } >
        <MemoryRouter>
            <RegisterScreen />
        </MemoryRouter>
    </Provider>
);



describe('Pruebas en <RegisterScreen />', () => {

    beforeEach(() => {
        jest.clearAllMocks();
    })


    test('debe mostrarse correctamente', () => {

        expect( wrapper ).toMatchSnapshot();
        
    });

    
    test('No hay registro si las contraseñas son diferentes', () => {

        wrapper.find('input[name="rPassword"]').simulate('change', {
            target: {
                name: 'rPassword',
                value: '123456',
            }
        });

        wrapper.find('input[name="rPassword2"]').simulate('change', {
            target: {
                name: 'rPassword2',
                value: '1234567',
            }
        });
        

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });

        expect( startRegister ).not.toHaveBeenCalled();
        expect( Swal.fire ).toHaveBeenCalledWith('Error', 'Las contraseñas no coinciden','error');
    
    })
    

    test('Registro con contraseñas iguales', () => {
        
        wrapper.find('input[name="rPassword"]').simulate('change', {
            target: {
                name: 'rPassword',
                value: 'hola mundo',
            }
        });

        wrapper.find('input[name="rPassword2"]').simulate('change', {
            target: {
                name: 'rPassword2',
                value: 'hola mundo',
            }
        });
        

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });

        expect( Swal.fire ).not.toHaveBeenCalled();
        // name, email, password
        expect(  startRegister ).toHaveBeenCalledWith("","","hola mundo");
    
    })
    

    
    
})
