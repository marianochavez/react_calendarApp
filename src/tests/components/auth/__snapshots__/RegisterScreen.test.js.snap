// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <RegisterScreen /> debe mostrarse correctamente 1`] = `
<section
  className="vh-100 gradient-custom"
>
  <div
    className="container py-5 h-100"
  >
    <div
      className="row d-flex justify-content-center align-items-center h-100"
    >
      <div
        className="col-12 col-md-8 col-lg-6 col-xl-5"
      >
        <div
          className="card card-login bg-dark text-white animate__animated animate__fadeIn animate__fast"
        >
          <div
            className="card-body p-5 text-center"
          >
            <div
              className="mb-md-5 mt-md-4 pb-4"
            >
              <h2
                className="fw-bold mb-2 text-uppercase"
              >
                Registro
              </h2>
              <p
                className="text-white-50 mb-5"
              >
                Por favor creese una cuenta!
              </p>
              <form
                onSubmit={[Function]}
              >
                <div
                  className="form-outline form-white mb-4"
                >
                  <input
                    autoComplete="off"
                    className="form-control form-control-lg"
                    name="rName"
                    onChange={[Function]}
                    placeholder="Nombre"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  className="form-outline form-white mb-4"
                >
                  <input
                    autoComplete="off"
                    className="form-control form-control-lg"
                    name="rEmail"
                    onChange={[Function]}
                    placeholder="Email"
                    type="email"
                    value=""
                  />
                </div>
                <div
                  className="form-outline form-white mb-4"
                >
                  <input
                    autoComplete="off"
                    className="form-control form-control-lg"
                    id="typePasswordX"
                    name="rPassword"
                    onChange={[Function]}
                    placeholder="Contraseña"
                    type="password"
                    value=""
                  />
                </div>
                <div
                  className="form-outline form-white mb-4"
                >
                  <input
                    autoComplete="off"
                    className="form-control form-control-lg"
                    id="typePasswordX2"
                    name="rPassword2"
                    onChange={[Function]}
                    placeholder="Confirmar contraseña"
                    type="password"
                    value=""
                  />
                </div>
                <button
                  className="btn btn-outline-light btn-lg px-5"
                  type="submit"
                >
                  Registrarme
                </button>
              </form>
            </div>
            <div>
              <p
                className="mb-0"
              >
                Ya estas registrado?
                 
                <a
                  className="text-white-50 fw-bold"
                  href="/login"
                  onClick={[Function]}
                >
                  Ingresar
                </a>
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;
